File: plugins/worktree-tools/skills/create-worktree/scripts/create_worktree.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import shlex
import shutil
import subprocess
import sys
from pathlib import Path


IGNORED_COPY_CANDIDATES = [".env", ".clubs-secrets", ".claude"]
BOOTSTRAP_COMMANDS = [
    "source ~/.nvm/nvm.sh && (nvm use 22.22.1 || nvm install 22.22.1)",
    "pnpm install",
    "pnpm rebuild @prisma/client prisma",
    "pnpm --filter api prisma:generate",
]


def run(cmd: str, cwd: Path, dry_run: bool) -> None:
    print(f"$ (cd {cwd} && {cmd})")
    if dry_run:
        return
    subprocess.run(["/bin/zsh", "-lc", cmd], cwd=cwd, check=True)


def git_root() -> Path:
    result = subprocess.run(
        ["git", "rev-parse", "--show-toplevel"],
        capture_output=True,
        text=True,
        check=True,
    )
    return Path(result.stdout.strip())


def primary_worktree_root(repo_root: Path) -> Path:
    result = subprocess.run(
        ["git", "worktree", "list", "--porcelain"],
        cwd=repo_root,
        capture_output=True,
        text=True,
        check=True,
    )
    for line in result.stdout.splitlines():
        if line.startswith("worktree "):
            return Path(line.removeprefix("worktree ").strip())
    raise RuntimeError("Could not determine the primary worktree root.")


def remote_branch_exists(repo_root: Path, branch: str) -> bool:
    result = subprocess.run(
        ["git", "show-ref", "--verify", "--quiet", f"refs/remotes/origin/{branch}"],
        cwd=repo_root,
    )
    return result.returncode == 0


def local_branch_exists(repo_root: Path, branch: str) -> bool:
    result = subprocess.run(
        ["git", "show-ref", "--verify", "--quiet", f"refs/heads/{branch}"],
        cwd=repo_root,
    )
    return result.returncode == 0


def copy_candidate(src_root: Path, dest_root: Path, relative: str, dry_run: bool) -> bool:
    src = src_root / relative
    dest = dest_root / relative
    if not src.exists():
        return False
    print(f"$ copy {src} -> {dest}")
    if dry_run:
        return True
    if src.is_dir():
        shutil.copytree(src, dest, dirs_exist_ok=True)
    else:
        dest.parent.mkdir(parents=True, exist_ok=True)
        shutil.copy2(src, dest)
    return True


def main() -> int:
    parser = argparse.ArgumentParser(
        description="Create and bootstrap a sibling worktree for ar-002-clubs."
    )
    parser.add_argument("--branch", required=True, help="Branch name to use in the worktree.")
    parser.add_argument(
        "--start-point",
        default="origin/dev",
        help="Git ref to branch from when creating a fresh branch. Default: origin/dev",
    )
    parser.add_argument(
        "--reuse-remote-branch",
        action="store_true",
        help="Reuse origin/<branch> if it already exists instead of branching from start-point.",
    )
    parser.add_argument(
        "--worktree-name",
        help="Override the target worktree directory name. Defaults to branch with '/' replaced by '-'.",
    )
    parser.add_argument(
        "--skip-bootstrap",
        action="store_true",
        help="Create and copy files only; skip install/rebuild/generate commands.",
    )
    parser.add_argument(
        "--dry-run",
        action="store_true",
        help="Print actions without changing git state or files.",
    )

    args = parser.parse_args()

    repo_root = git_root()
    source_root = primary_worktree_root(repo_root)
    worktrees_root = source_root.parent / "clubs-worktrees"
    worktree_name = args.worktree_name or args.branch.replace("/", "-")
    target = worktrees_root / worktree_name

    print(f"Current repo root: {repo_root}")
    print(f"Source repo root: {source_root}")
    print(f"Worktrees root: {worktrees_root}")
    print(f"Target branch: {args.branch}")
    print(f"Target worktree: {target}")

    if target.exists():
        raise SystemExit(f"Target worktree already exists: {target}")

    if args.dry_run:
        print(f"$ mkdir -p {worktrees_root}")
    else:
        worktrees_root.mkdir(parents=True, exist_ok=True)

    if args.reuse_remote_branch and remote_branch_exists(source_root, args.branch):
        if local_branch_exists(source_root, args.branch):
            cmd = ["git", "worktree", "add", str(target), args.branch]
        else:
            cmd = ["git", "worktree", "add", "-b", args.branch, str(target), f"origin/{args.branch}"]
    else:
        cmd = ["git", "worktree", "add", "-b", args.branch, str(target), args.start_point]

    print("$ " + " ".join(shlex.quote(part) for part in cmd))
    if not args.dry_run:
        subprocess.run(cmd, cwd=source_root, check=True)

    copied = []
    for candidate in IGNORED_COPY_CANDIDATES:
        if copy_candidate(source_root, target, candidate, args.dry_run):
            copied.append(candidate)

    if copied:
        print("Copied ignored files:")
        for item in copied:
            print(f" - {item}")
    else:
        print("No ignored files were copied.")

    if not args.skip_bootstrap:
        for command in BOOTSTRAP_COMMANDS:
            run(command, target, args.dry_run)

    print("Done.")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: plugins/worktree-tools/skills/clubs-task-start/scripts/create_worktree.mjs
#!/usr/bin/env node

import { cpSync, existsSync, mkdirSync, readFileSync } from "node:fs";
import { dirname, join, resolve } from "node:path";
import { execFileSync, spawnSync } from "node:child_process";

const IGNORED_COPY_CANDIDATES = [".env", ".clubs-secrets", ".claude"];
const BOOTSTRAP_COMMANDS = [
  "pnpm install",
  "pnpm rebuild @prisma/client prisma",
  "pnpm --filter api prisma:generate",
];

function parseArgs(argv) {
  const options = {
    branch: "",
    startPoint: "origin/dev",
    reuseRemoteBranch: false,
    worktreeName: "",
    skipBootstrap: false,
    dryRun: false,
    help: false,
  };

  for (let index = 0; index < argv.length; index += 1) {
    const arg = argv[index];

    if (arg === "--") {
      continue;
    }

    if (arg === "-h" || arg === "--help") {
      options.help = true;
      continue;
    }

    if (arg === "--reuse-remote-branch") {
      options.reuseRemoteBranch = true;
      continue;
    }

    if (arg === "--skip-bootstrap") {
      options.skipBootstrap = true;
      continue;
    }

    if (arg === "--dry-run") {
      options.dryRun = true;
      continue;
    }

    if (arg === "--branch") {
      options.branch = argv[index + 1] ?? "";
      index += 1;
      continue;
    }

    if (arg === "--start-point") {
      options.startPoint = argv[index + 1] ?? "";
      index += 1;
      continue;
    }

    if (arg === "--worktree-name") {
      options.worktreeName = argv[index + 1] ?? "";
      index += 1;
      continue;
    }

    throw new Error(`Unknown argument: ${arg}`);
  }

  return options;
}

function printHelp() {
  console.log(`Usage: pnpm clubs-task-start -- --branch <name> [options]

Options:
  --branch <name>             Branch name to use in the worktree
  --start-point <ref>         Base ref for a fresh branch (default: origin/dev)
  --reuse-remote-branch       Reuse origin/<branch> when it already exists
  --worktree-name <name>      Override the target directory name
  --skip-bootstrap            Skip install/rebuild/generate commands
  --dry-run                   Print actions without changing files
  -h, --help                  Show this help message
`);
}

function runGit(args, cwd) {
  return execFileSync("git", args, {
    cwd,
    encoding: "utf8",
  }).trim();
}

function gitRoot() {
  return runGit(["rev-parse", "--show-toplevel"], process.cwd());
}

function primaryWorktreeRoot(repoRoot) {
  const output = runGit(["worktree", "list", "--porcelain"], repoRoot);
  const line = output.split("\n").find(entry => entry.startsWith("worktree "));

  if (!line) {
    throw new Error("Could not determine the primary worktree root.");
  }

  return line.replace("worktree ", "").trim();
}

function remoteBranchExists(repoRoot, branch) {
  const result = spawnSync(
    "git",
    ["show-ref", "--verify", "--quiet", `refs/remotes/origin/${branch}`],
    { cwd: repoRoot },
  );
  return result.status === 0;
}

function localBranchExists(repoRoot, branch) {
  const result = spawnSync(
    "git",
    ["show-ref", "--verify", "--quiet", `refs/heads/${branch}`],
    { cwd: repoRoot },
  );
  return result.status === 0;
}

function normalizeBranch(input) {
  if (!input) return "";
  if (/^TU-\d+$/i.test(input)) {
    return input.toUpperCase();
  }
  if (/^\d+$/.test(input)) {
    return `TU-${input}`;
  }
  return input;
}

function currentBranch(repoRoot) {
  return runGit(["branch", "--show-current"], repoRoot);
}

function isWorktreeClean(repoRoot) {
  return runGit(["status", "--short"], repoRoot) === "";
}

function runGitCommand(args, cwd, dryRun) {
  console.log(`$ git ${args.join(" ")}`);
  if (dryRun) {
    return;
  }
  const result = spawnSync("git", args, {
    cwd,
    stdio: "inherit",
  });
  if (result.status !== 0) {
    throw new Error(`git ${args[0]} failed`);
  }
}

function readNodeVersion(repoRoot) {
  const nvmrcPath = join(repoRoot, ".nvmrc");
  if (!existsSync(nvmrcPath)) {
    return "";
  }

  return readFileSync(nvmrcPath, "utf8").trim();
}

function withNodeVersion(command, nodeVersion) {
  if (process.platform === "win32" || !nodeVersion) {
    return command;
  }

  return [
    'export NVM_DIR="${NVM_DIR:-$HOME/.nvm}"',
    '[ -s "$NVM_DIR/nvm.sh" ]',
    '. "$NVM_DIR/nvm.sh"',
    `nvm use ${nodeVersion} >/dev/null || nvm install ${nodeVersion} >/dev/null`,
    command,
  ].join(" && ");
}

function runShell(command, cwd, dryRun, nodeVersion = "") {
  console.log(`$ (cd ${cwd} && ${command})`);
  if (dryRun) return;
  const shellCommand =
    process.platform === "win32"
      ? { file: "cmd.exe", args: ["/c", withNodeVersion(command, nodeVersion)] }
      : {
          file: process.env.SHELL || "/bin/bash",
          args: ["-lc", withNodeVersion(command, nodeVersion)],
        };
  const result = spawnSync(shellCommand.file, shellCommand.args, {
    cwd,
    stdio: "inherit",
  });
  if (result.status !== 0) {
    throw new Error(`Command failed: ${command}`);
  }
}

function copyCandidate(sourceRoot, targetRoot, relativePath, dryRun) {
  const source = join(sourceRoot, relativePath);
  const target = join(targetRoot, relativePath);

  if (!existsSync(source)) {
    return false;
  }

  console.log(`$ copy ${source} -> ${target}`);
  if (dryRun) {
    return true;
  }

  mkdirSync(dirname(target), { recursive: true });
  cpSync(source, target, { recursive: true, force: true });
  return true;
}

function main() {
  const options = parseArgs(process.argv.slice(2));
  if (options.help) {
    printHelp();
    return;
  }

  if (!options.branch) {
    throw new Error("--branch is required");
  }

  options.branch = normalizeBranch(options.branch);

  const repoRoot = gitRoot();
  const sourceRoot = primaryWorktreeRoot(repoRoot);
  const nodeVersion = readNodeVersion(sourceRoot);
  const worktreesRoot = resolve(sourceRoot, "..", "clubs-worktrees");
  const worktreeName =
    options.worktreeName || options.branch.replaceAll("/", "-");
  const target = join(worktreesRoot, worktreeName);

  console.log(`Current repo root: ${repoRoot}`);
  console.log(`Source repo root: ${sourceRoot}`);
  console.log(`Worktrees root: ${worktreesRoot}`);
  console.log(`Target branch: ${options.branch}`);
  console.log(`Target worktree: ${target}`);

  if (existsSync(target)) {
    throw new Error(`Target worktree already exists: ${target}`);
  }

  if (options.dryRun) {
    console.log(`$ mkdir -p ${worktreesRoot}`);
  } else {
    mkdirSync(worktreesRoot, { recursive: true });
  }

  const shouldCreateFreshBranch =
    !options.reuseRemoteBranch ||
    !remoteBranchExists(sourceRoot, options.branch);

  if (shouldCreateFreshBranch) {
    const sourceBranch = currentBranch(sourceRoot);

    if (sourceBranch === "dev") {
      runGitCommand(["pull", "origin", "dev"], sourceRoot, options.dryRun);
    } else {
      if (!isWorktreeClean(sourceRoot)) {
        throw new Error(
          "Primary repo worktree must be clean before syncing dev for a fresh branch. Switch or clean the primary repo first.",
        );
      }

      runGitCommand(["switch", "dev"], sourceRoot, options.dryRun);
      runGitCommand(["pull", "origin", "dev"], sourceRoot, options.dryRun);
      runGitCommand(["switch", sourceBranch], sourceRoot, options.dryRun);
    }
  }

  let worktreeArgs;
  if (
    options.reuseRemoteBranch &&
    remoteBranchExists(sourceRoot, options.branch)
  ) {
    if (localBranchExists(sourceRoot, options.branch)) {
      worktreeArgs = ["worktree", "add", target, options.branch];
    } else {
      worktreeArgs = [
        "worktree",
        "add",
        "-b",
        options.branch,
        target,
        `origin/${options.branch}`,
      ];
    }
  } else {
    worktreeArgs = [
      "worktree",
      "add",
      "-b",
      options.branch,
      target,
      options.startPoint,
    ];
  }

  console.log(`$ git ${worktreeArgs.join(" ")}`);
  if (!options.dryRun) {
    const result = spawnSync("git", worktreeArgs, {
      cwd: sourceRoot,
      stdio: "inherit",
    });
    if (result.status !== 0) {
      throw new Error("git worktree add failed");
    }
  }

  const copied = [];
  for (const candidate of IGNORED_COPY_CANDIDATES) {
    if (copyCandidate(sourceRoot, target, candidate, options.dryRun)) {
      copied.push(candidate);
    }
  }

  if (copied.length > 0) {
    console.log("Copied ignored files:");
    for (const item of copied) {
      console.log(` - ${item}`);
    }
  } else {
    console.log("No ignored files were copied.");
  }

  if (!options.skipBootstrap) {
    for (const command of BOOTSTRAP_COMMANDS) {
      runShell(command, target, options.dryRun, nodeVersion);
    }
  }

  console.log("Done.");
}

try {
  main();
} catch (error) {
  console.error(error instanceof Error ? error.message : String(error));
  process.exit(1);
}
